core/matrix/ell.cpp

```diff
@@ -101,39 +102,29 @@ size_type calculate_max_nnz_per_row(
 template <typename ValueType, typename IndexType>
 void Ell<ValueType, IndexType>::apply_impl(const LinOp *b, LinOp *x) const
 {
-    using ComplexDense = Dense<to_complex<ValueType>>;
-
-    if (dynamic_cast<const Dense<ValueType> *>(b)) {
-        this->get_executor()->run(ell::make_spmv(this, as<Dense<ValueType>>(b),
-                                                 as<Dense<ValueType>>(x)));
-    } else {
-        auto dense_b = as<ComplexDense>(b);
-        auto dense_x = as<ComplexDense>(x);
-        this->apply(dense_b->create_real_view().get(),
-                    dense_x->create_real_view().get());
-    }
+    mixed_precision_dispatch_spmv<ValueType>(
+        [&](auto dense_b, auto dense_x) {
+            this->get_executor()->run(ell::make_spmv(this, dense_b, dense_x));
+        },
+        b, x);
 }
 
 
 template <typename ValueType, typename IndexType>
 void Ell<ValueType, IndexType>::apply_impl(const LinOp *alpha, const LinOp *b,
                                            const LinOp *beta, LinOp *x) const
 {
-    using ComplexDense = Dense<to_complex<ValueType>>;
-    using RealDense = Dense<remove_complex<ValueType>>;
-
-    if (dynamic_cast<const Dense<ValueType> *>(b)) {
-        this->get_executor()->run(ell::make_advanced_spmv(
-            as<Dense<ValueType>>(alpha), this, as<Dense<ValueType>>(b),
-            as<Dense<ValueType>>(beta), as<Dense<ValueType>>(x)));
-    } else {
-        auto dense_b = as<ComplexDense>(b);
-        auto dense_x = as<ComplexDense>(x);
-        auto dense_alpha = as<RealDense>(alpha);
-        auto dense_beta = as<RealDense>(beta);
-        this->apply(dense_alpha, dense_b->create_real_view().get(), dense_beta,
-                    dense_x->create_real_view().get());
-    }
+    mixed_precision_dispatch_spmv<ValueType>(
+        [&](auto dense_b, auto dense_x) {
+            auto converted_alpha = make_temporary_conversion<ValueType>(alpha);
+            auto converted_beta =
+                make_temporary_conversion<typename std::remove_reference_t<
+                    decltype(*dense_x)>::value_type>(beta);
+            this->get_executor()->run(
+                ell::make_advanced_spmv(converted_alpha.get(), this, dense_b,
+                                        converted_beta.get(), dense_x));
+        },
+        b, x);
```

Comment:
Oh right, this question was actually for me (and is still relevant for the current state, even though the above code is outdated).

We are doing two different things here:
1. `dense_b` and `dense_x` are `x` and `b` cast "from their dynamic type to their static type", i.e. if `x` is `Dense<float>`, then the lambda instantiation with `Dense<float> dense_x` parameter type is called.
2. Due to their small size (and not wanting to have 4x more parameter type combinations), the scalars are converted to the correct type instead of using the above procedure. So they do indeed need to be treated differently for this to work and make sense.